### DISCUSSION TITLE
Flux pattern에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by greenblues1190** [woowacourse-study/2022-woowahan-bansanghwe#112] [React]
Flux 패턴은 페이스북이 2014년 발표한 프론트엔드 아키텍쳐 패턴입니다. 기존 MVC(와 MV*) 프레임워크의 양방향 데이터 흐름의 단점을 극복하고 보다 확장성있는 웹 어플리케이션을 개발하기 위해 등장했습니다.

## MV\* vs Flux

MV\*와 Flux 패턴의 가장 큰 차이는 단방향 데이터 흐름입니다. Ember.js, Angular1 등 MV\* 프레임워크는 양방향 데이터 바인딩을 지원해 뷰와 모델이 서로를 업데이트할 수 있습니다. 복잡한 어플리케이션에서 이러한 구조를 사용하면 데이터의 흐름이 복잡해지기 때문에 확장과 디버깅이 어렵습니다.

### 액션

이벤트의 정보를 담은 자바스크립트 객체입니다.

### 디스패쳐

이벤트를 각 스토어로 전달합니다.

### 스토어

컴포넌트를 위한 복수의 데이터 저장소입니다.

### 뷰

상태가 변경되면 뷰를 새로 그립니다.(ex.리액트) 상태를 받아 모든 뷰를 새로 그리는 간단한 일을 하기 때문에 전체 시스템의 데이터 흐름을 단순화할 수 있습니다. 모든 뷰를 새로 그리는 과정에서 많은 연산이 발생하기에 리액트처럼 비교 알고리즘을 통해 변경된 뷰만 렌더링하는 경우가 많습니다. Flux와 리액트는 땔래야 땔 수 없는 관계라 할 수 있겠죠?

## MVC는 양방향 데이터 바인딩이다?

순수한 MVC는 단방향 데이터 바인딩입니다. 웹 어플리케이션 프레임워크에서 MVP, MVVM과 같은 MV* 아키텍쳐를 MVC로 부르던 것이 2014년 이전의 페이스북 아키텍쳐와 같은 참사를 낳았습니다. Flux는 이벤트 드리븐을 적용하고 이름을 새로 붙인 MVC라고 할 수 있습니다.

## Redux

Flux 패턴의 개념을 실현 혹은 차용한 여러 구현체들이 있는데요, Redux만 살펴보겠습니다.

Redux는 단일 스토어에서 상태를 관리합니다. 모든 상태는 순수 함수인 리듀서를 통해 변경됩니다. 단일 저장소에서 모든 상태 변경이 이루어지기 때문에 강력한 시간여행형 디버깅을 지원합니다.

## 상태 정규화

스토어의 상태는 ORM 스타일이 아니라 여러 형태의 상태를 저장할 수 있기 때문에 중첩, 관계형 데이터를 처리할 때 데이터 구조가 복잡하고 중복될 수 있습니다. 이를 위해 스토어의 일부를 데이터베이스처럼 취급하여 데이터를 구성하는 것을 상태 정규화라고 합니다. 리덕스의 상태 정규화 예시를 볼까요?

### 정규화된 상태 설계

- 각 도메인은 상태 내에서 자신만의 테이블을 가집니다.
- 각 테이블은 엔티티를 객체 형태로 저장합니다. 각 엔티티는 id를 통해 조회가능합니다.
- 각 도메인은 엔티티들의 순서를 나타내는 배열을 가집니다.
- 각 엔티티에 대한 참조는 해당 엔티티의 id로 이루어져야 합니다.

이렇게 정규화된 상태는 깊이가 줄어들고 중복이 제거되어 한 곳에서만 변경을 처리할 수 있게 됩니다. 또한 id를 가지기 때문에 조회 로직이 간단해집니다.

**Comment by al-bur** [comment]
flux pattern이 무엇인가요?

**Comment by al-bur** [comment]
왜 프로젝트에서 redux를 사용했나요?

**Comment by al-bur** [comment]
디버깅 용으로 사용하신건가요??

**Comment by al-bur** [comment]
context api말고 redux를 사용한 이유는?

**Comment by al-bur** [comment]
redux의 단점은?

**Comment by al-bur** [comment]
앞으로 프로젝트 하실 때 상태관리를 어떻게 할 건지?

**Comment by al-bur** [comment]
왜 reducer가 순수함수여야하나요?

**Reply by al-bur** [reply]
[테스트에 용이하다고 해서 꼬리질문] 직접 테스트 해봤나요?

**Comment by airman5573** [comment]
페이스북에서 어떤 이유로 Flux Pattern을 적용한건가요?

**Comment by airman5573** [comment]
뷰와 상태의 복잡성을 Flux가 어떻게 해결한건가요?

**Comment by airman5573** [comment]
선언형과 명령형의 차이점에 대해 설명해 주세요

**Comment by airman5573** [comment]
리덕스의 단점은 무엇이라고 생각하시나요?

**Comment by airman5573** [comment]
왜 리듀서는 순수 함수여야 하나요?

**Comment by airman5573** [comment]
react-redux는 왜 필요한가요?
